src/functional.ts

```diff
@@ -136,6 +179,103 @@ export function functionalInterfaceFactory({
+  }
+
+  function convertToBlstSignatureArg(signature: SignatureArg): blst.SignatureArg {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    return signature instanceof Signature ? ((signature as any).value as blst.Signature) : signature;
+  }
+
+  /**
+   * Verifies if signature is message signed with given public key.
+   * @param publicKey
+   * @param message
+   * @param signature
+   */
+  async function asyncVerify(publicKey: PublicKeyArg, message: Uint8Array, signature: SignatureArg): Promise<boolean> {
+    if (implementation === "herumi") return verify(publicKey, message, signature);
+    try {
+      return blst.asyncVerify(message, convertToBlstPublicKeyArg(publicKey), convertToBlstSignatureArg(signature));
+    } catch {
+      return false;
+    }
+  }
+
+  /**
+   * Verifies if aggregated signature is same message signed with given public keys.
+   * @param publicKeys
+   * @param message
+   * @param signature
+   */
+  async function asyncVerifyAggregate(
+    publicKeys: PublicKeyArg[],
+    message: Uint8Array,
+    signature: SignatureArg
+  ): Promise<boolean> {
+    if (implementation === "herumi") return verifyAggregate(publicKeys, message, signature);
+    try {
+      return blst.asyncFastAggregateVerify(
+        message,
+        publicKeys.map((key) => convertToBlstPublicKeyArg(key)),
+        convertToBlstSignatureArg(signature)
+      );
+    } catch {
+      return false;
+    }
+  }
+
+  /**
+   * Verifies if signature is list of message signed with corresponding public key.
+   * @param publicKeys
+   * @param messages
+   * @param signature
+   * @param fast Check if all messages are different
+   */
+  async function asyncVerifyMultiple(
+    publicKeys: PublicKeyArg[],
+    messages: Uint8Array[],
+    signature: SignatureArg
+  ): Promise<boolean> {
+    if (implementation === "herumi") return verifyMultiple(publicKeys, messages, signature);
+    try {
+      return blst.asyncAggregateVerify(
+        messages,
+        publicKeys.map((key) => convertToBlstPublicKeyArg(key)),
+        convertToBlstSignatureArg(signature)
+      );
+    } catch {
+      return false;
+    }
+  }
+
+  /**
+   * Verifies multiple signatures at once returning true if all valid or false
+   * if at least one is not. Optimization useful when knowing which signature is
+   * wrong is not relevant, i.e. verifying an entire Eth2.0 block.
+   *
+   * This method provides a safe way to do so by multiplying each signature by
+   * a random number so an attacker cannot craft a malicious signature that won't
+   * verify on its own but will if it's added to a specific predictable signature
+   * https://ethresear.ch/t/fast-verification-of-multiple-bls-signatures/5407
+   */
+  async function asyncVerifyMultipleSignatures(sets: SignatureSet[]): Promise<boolean> {
+    if (implementation === "herumi") return verifyMultipleSignatures(sets);
+    try {
+      return blst.asyncVerifyMultipleAggregateSignatures(
```

Comment:
I'm curious why the name has "aggregate" there, as I understand these are separate signature sets

Reply:
Thats a good questions... It was how it was named in blst when I found that project and just left the names as is.  I suppose they are all aggregated onto a single point context and return a single bool.  So I guess the verifications are aggregated but technically the signatures are not... The functions in `supranational/blst` is [mul_n_aggregate](https://github.com/supranational/blst/blob/704c7f6d5f99ebb6bda84f635122e449ee51aa48/bindings/blst.hpp#L919) so perhaps that is why.

Reply:
guessing `multipleAggregateVerifications` is "more correct" though in reality